_source/_posts/2022-07-14-add-auth-to-any-app-with-oauth2-proxy.adoc

```diff
@@ -0,0 +1,583 @@
+
+====
+[source,conf]
+----
+server {
+    listen 80;
+    server_name _;
+
+    location = /oauth2/auth {
+        internal; # <.>
+        proxy_pass       http://oauth2-proxy:4180;
+        proxy_set_header Host             $host;
+        proxy_set_header X-Real-IP        $remote_addr;
+        proxy_set_header X-Scheme         $scheme;
+        # nginx auth_request includes headers but not body
+        proxy_set_header Content-Length   "";
+        proxy_pass_request_body           off;
+    }
+
+    location / {
+        auth_request /oauth2/auth; # <.>
+
+        auth_request_set $email  $upstream_http_x_auth_request_email; # <.>
+        proxy_set_header X-Email $email;
+        auth_request_set $user  $upstream_http_x_auth_request_user;
+        proxy_set_header X-User  $user;
+        auth_request_set $token  $upstream_http_x_auth_request_access_token;
+        proxy_set_header X-Access-Token $token;
+        auth_request_set $auth_cookie $upstream_http_set_cookie;
+        add_header Set-Cookie $auth_cookie;
+
+        proxy_set_header Host $host;
+        proxy_set_header X-Real-IP $remote_addr;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header X-Forwarded-Host $host:80;
+        proxy_set_header X-Forwarded-Port 80;
+        proxy_set_header X-Forwarded-Server $host;
+        proxy_set_header X-Forwarded-Proto http;
+
+        proxy_http_version 1.1; # <.>
+        proxy_pass http://web-app:8080/; # <.>
+    }
+}
+----
+
+<.> Don't expose this route to external clients.
+<.> Nginx makes a request to `oauth2-proxy` REST API, to verify the auth for this request.
+<.> Add headers returned from the auth request.
+<.> HTTP 1.0 is the default if not set.
+<.> Send authenticated requests to the web app.
+====
+
+Restart the docker containers and verify everything is working with HTTPie (make sure you are using port `80` now).
+
+[source,bash]
+----
+http localhost/echo "Authorization: Bearer ${TOKEN}"
+----
+
+If you remove or change the `Authorization` header, a `401` will be returned. Browser requests will now also return a `401`!
+
+Almost done! We still need to make everything work for both API clients and browsers, and handle sign-out requests.
+
+== Route  all traffic through Nginx
+
+Sending all the traffic through Nginx has the added benefit of giving you control of how the OAuth2 Proxy endpoints are exposed. For example, the previous section marked the `/oauth2/auth` route as "internal," so only the `auth_requst` directive can use it.
+
+In the `nginx-default.conf.template`, add a couple of new `location` sections to expose the other `/oauth2` endpoints. The first `location` will handle the OAuth 2.0-related requests like the redirect callback. The second will configure the sign-out endpoint to accept only POST requests. (This prevents a rogue GET request from ending the user's session).
```

Comment:
```suggestion
In the `nginx-default.conf.template`, add a couple of new `location` sections to expose the other `/oauth2` endpoints. The first `location` will handle the OAuth 2.0-related requests like the redirect callback. The second will configure the sign-out endpoint to accept only POST requests. (This prevents a rogue GET request from ending the user's session.)
```
punctuation tweak. if the sentence btwn the parens is complete, the period goes on the inside.